File: evaluation/S0/c/Easy/23-balanced-binary-tree.c
// Source: https://leetcode.com/problems/balanced-binary-tree/   |   Difficulty: Easy
//
// Problem Description:
// Given a binary tree, determine if it is height-balanced.
//
// Example:
// Input: root = [3,9,20,null,null,15,7]
// Output: true
//
// Constraints:
// The number of nodes in the tree is in the range [0, 5000].
// 	-104 <= Node.val <= 104
//

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool isBalanced(struct TreeNode* root) {
    bool balanced = true;

    int height(struct TreeNode* node) {
        if (!node || !balanced) {
            return 0;
        }

        int leftHeight = height(node->left);
        int rightHeight = height(node->right);

        if (abs(leftHeight - rightHeight) > 1) {
            balanced = false;
        }

        return 1 + (leftHeight > rightHeight ? leftHeight : rightHeight);
    }

    height(root);
    return balanced;        
}

File: evaluation/S0/javascript/Easy/23-balanced-binary-tree.js
// Source: https://leetcode.com/problems/balanced-binary-tree/   |   Difficulty: Easy
//
// Problem Description:
// Given a binary tree, determine if it is height-balanced.
//
// Example:
// Input: root = [3,9,20,null,null,15,7]
// Output: true
//
// Constraints:
// The number of nodes in the tree is in the range [0, 5000].
// 	-104 <= Node.val <= 104
//

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
    const checkBalance = (node) => {
        if (!node) return 0;
        const leftHeight = checkBalance(node.left);
        if (leftHeight === -1) return -1;
        const rightHeight = checkBalance(node.right);
        if (rightHeight === -1) return -1;
        if (Math.abs(leftHeight - rightHeight) > 1) return -1;
        return Math.max(leftHeight, rightHeight) + 1;
    };
    return checkBalance(root) !== -1;
};
